src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsClient/DiagnosticsClient.cs

```diff
@@ -169,15 +175,70 @@ public async Task WriteDumpAsync(DumpType dumpType, string dumpPath, bool logDum
-            if (!ValidateResponseMessage(response, nameof(WriteDumpAsync), ValidateResponseOptions.UnknownCommandReturnsFalse))
+            if (!ValidateWriteDumpResponseMessage(response, errorMessage: true))
             {
-                if ((flags & ~WriteDumpFlags.LoggingEnabled) != 0)
+                request = CreateWriteDumpMessage(DumpCommandId.GenerateCoreDump2, dumpType, dumpPath, flags);
+                response = await IpcClient.SendMessageAsync(_endpoint, request, token).ConfigureAwait(false);
+                if (!ValidateWriteDumpResponseMessage(response, errorMessage: false))
                 {
-                    throw new ArgumentException($"Only {nameof(WriteDumpFlags.LoggingEnabled)} flag is supported by this runtime version", nameof(flags));
+                    if ((flags & ~WriteDumpFlags.LoggingEnabled) != 0)
+                    {
+                        throw new ArgumentException($"Only {nameof(WriteDumpFlags.LoggingEnabled)} flag is supported by this runtime version", nameof(flags));
+                    }
+                    request = CreateWriteDumpMessage(dumpType, dumpPath, logDumpGeneration: (flags & WriteDumpFlags.LoggingEnabled) != 0);
+                    response = await IpcClient.SendMessageAsync(_endpoint, request, token).ConfigureAwait(false);
+                    if (!ValidateWriteDumpResponseMessage(response, errorMessage: false))
+                    {
+                        throw new UnsupportedCommandException($"Write dump failed - command is not supported.");
+                    }
                 }
-                await WriteDumpAsync(dumpType, dumpPath, logDumpGeneration: (flags & WriteDumpFlags.LoggingEnabled) != 0, token);
+            }
+        }
+
+        private static bool ValidateWriteDumpResponseMessage(IpcMessage response, bool errorMessage)
+        {
+            switch ((DiagnosticsServerResponseId)response.Header.CommandId)
+            {
+                case DiagnosticsServerResponseId.OK:
+                    return true;
+
+                case DiagnosticsServerResponseId.Error:
+                    int index = 0;
+                    uint hr = BitConverter.ToUInt32(response.Payload, index);
+                    index += sizeof(UInt32);
+                    string message = null;
+                    switch (hr)
+                    {
+                        case (uint)DiagnosticsIpcError.UnknownCommand:
+                            return false;
+
+                        case (uint)DiagnosticsIpcError.InvalidArgument:
+                            throw new UnsupportedCommandException("Write dump failed - Invalid command argument.");
+
+                        case (uint)DiagnosticsIpcError.NotSupported:
+                            throw new NotSupportedException("Write dump failed - Not supported by this runtime.");
+
+                        // Only this error code (E_FAIL) can have a error message
+                        case (uint)DiagnosticsIpcError.Fail:
```

Comment:
Other than the index tracking and custom error message, this seems to be the only difference compared to ValidateResponseMessage. Consider folding this back into ValidateResponseMessage and adding another ValidateResponseOption such as FailHasErrorMessage. This keeps all of the response processing consistent, prevents different implementations from diverging unnecessarily, and allows other future commands to opt into this functionality easily without reimplementation.